src/Native/Unix/System.Native/pal_console.c

```diff
@@ -71,92 +73,200 @@ void SystemNative_SetKeypadXmit(const char* terminfoString)
-{
-    struct termios newTermios;
-    if (tcgetattr(STDIN_FILENO, &newTermios) >= 0)
+    if (!forChild)
+    {
+        termios.c_iflag &= (uint32_t)(~(IXON | IXOFF));
+        termios.c_lflag &= (uint32_t)(~(ECHO | ICANON | IEXTEN));
+    }
+
+    termios.c_cc[VMIN] = minChars;
+    termios.c_cc[VTIME] = decisecondsTimeout;
+
+    // Check if the settings have changed.
+    if (g_hasCurrentTermios)
     {
-        if (!g_readInProgress)
+        if (g_currentTermios.c_lflag == termios.c_lflag &&
+            g_currentTermios.c_iflag == termios.c_iflag &&
+            g_currentTermios.c_cc[VMIN] == termios.c_cc[VMIN] &&
+            g_currentTermios.c_cc[VMIN] == termios.c_cc[VMIN])
         {
-            // Store the original settings, but only if we didn't already.  This function
-            // may be called when the process is resumed after being suspended, and if
-            // that happens during a read, we'll call this function to reset the attrs.
-            g_preReadTermios = newTermios;
+            return true;
         }
+    }
+
+    return TcSetAttr(&termios, blockIfBackground);
+}
 
-        newTermios.c_iflag &= (uint32_t)(~(IXON | IXOFF));
-        newTermios.c_lflag &= (uint32_t)(~(ECHO | ICANON | IEXTEN));
-        newTermios.c_cc[VMIN] = minChars;
-        newTermios.c_cc[VTIME] = decisecondsTimeout;
-        IncorporateBreak(&newTermios, g_signalForBreak);
+void UninitializeTerminal()
+{
+    assert(g_haveInitTermios);
+
+    // This method is called on SIGQUIT/SIGINT from the signal dispatching thread
+    // and on atexit.
 
-        if (tcsetattr(STDIN_FILENO, TCSANOW, &newTermios) >= 0)
+    if (pthread_mutex_lock(&g_lock) == 0)
+    {
+        if (!g_terminalUninitialized)
         {
-            g_currTermios = newTermios;
-            g_readInProgress = true;
+            TcSetAttr(&g_initTermios, /* blockIfBackground */ false);
+
+            g_terminalUninitialized = true;
         }
+
+        pthread_mutex_unlock(&g_lock);
+    }
+}
+
+void SystemNative_InitializeConsoleBeforeRead(uint8_t minChars, uint8_t decisecondsTimeout)
+{
+    if (pthread_mutex_lock(&g_lock) == 0)
+    {
+        g_reading = true;
+
+        ConfigureTerminal(g_signalForBreak, /* forChild */ false, minChars, decisecondsTimeout, /* blockIfBackground */ true);
+
+        pthread_mutex_unlock(&g_lock);
     }
 }
 
 void SystemNative_UninitializeConsoleAfterRead()
 {
-    if (g_readInProgress)
+    if (pthread_mutex_lock(&g_lock) == 0)
     {
-        g_readInProgress = false;
-
-        int tmpErrno = errno; // preserve any errors from before uninitializing
-        IncorporateBreak(&g_preReadTermios, g_signalForBreak);
-        int ret = tcsetattr(STDIN_FILENO, TCSANOW, &g_preReadTermios);
-        assert(ret >= 0); // shouldn't fail, but if it does we don't want to fail in release
-        (void)ret;
-        errno = tmpErrno;
+        g_reading = false;
```

Comment:
Why don't we need to undo any of the work done by ConfigureTerminal, in particular in the case where there are child processes?

Reply:
As long as a read is in progress, we don't configure the terminal for child processes (g_reading check). Because we are reading we assume our children aren't using the terminal, so at the end of a read, we can keep the terminal as-is.

Reply:
I'm still not quite following.

Regardless of whether there are child processes, ConfigureTerminal modifies VMIN and VTIME.  Those don't need to be reset?

SystemNative_InitializeConsoleBeforeRead calls ConfigureTerminal with `/* forChild */ false`, and if forChild is false, ConfigureTerminal sets c_lflag and c_lflag.

And what happens if we launch a child process that might interact with the console, and then we do a read?  Won't we update the terminal then?

I'm obviously missing something.  Is there a summary somewhere that outlines the scheme being used for all of this?

Reply:
> ConfigureTerminal modifies VMIN and VTIME. Those don't need to be reset?

For reads made by the .NET applications, we set those before making the `read` call in `InitializeConsoleBeforeRead`.

> SystemNative_InitializeConsoleBeforeRead calls ConfigureTerminal with /* forChild */ false, and if forChild is false, ConfigureTerminal sets c_lflag and c_lflag. Those don't need to be reset?

If a child is launched (and the .NET application isn't reading) we reset those to the initial terminal settings.

> And what happens if we launch a child process that might interact with the console, and then we do a read? Won't we update the terminal then?

Because ProcessStartInfo defaults grant access to the terminal. So looking at those it is not clear if access is given to the terminal intentionally.
When we see a Console.Read, we assume our application is supposed to use the terminal (and not the child), and we reconfigure for our reading.

> I'm obviously missing something. Is there a summary somewhere that outlines the scheme being used for all of this?

Maybe to be extended...

https://github.com/dotnet/corefx/blob/8065fe74055436cb8b52b25822ab5dd95a3d9b6a/src/Native/Unix/System.Native/pal_console.c#L86-L91

Reply:
> For reads made by the .NET applications, we set those before making the read call in InitializeConsoleBeforeRead.

What about P/Invokes from the .NET application?  Is that too obscure?

> If a child is launched (and the .NET application isn't reading) we reset those to the initial terminal settings.

And if the app is reading it doesn't matter because...?

> When we see a Console.Read, we assume our application is supposed to use the terminal (and not the child), and we reconfigure for our reading.

What happens if the child process then tries to read?  One of the two ends up suspending, right?  And then I guess upon resumption the terminal is reconfigured?

Reply:
> What about P/Invokes from the .NET application? Is that too obscure?

If those are mixed and matched with Console operations then any state we track can become incorrect.

> And if the app is reading it doesn't matter because...?

If the app is reading, we assume the child won't be really using the terminal. In more detail: https://github.com/dotnet/corefx/pull/35621#issuecomment-475501794

> What happens if the child process then tries to read? One of the two ends up suspending, right? And then I guess upon resumption the terminal is reconfigured?

Suspending happens when doing a read while in background. Multiple processes can read simultaneously and they will get bits and pieces of available input. If two applications are reading from terminal input at the same time, there is a usability/design issue with the application.

This is why I think it is ok to assume that only one process is using the terminal interactively.